src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs

```diff
@@ -760,30 +911,63 @@ async Task Core(MultiConnectSocketAsyncEventArgs internalArgs, Task<IPAddress[]>
                         // the same socket handle can't be used for another connect, so we swap in a new handle under the covers if
                         // possible.  We do this not just for the 2nd+ address but also for the first in case the Socket was already
                         // used for a connection attempt outside of this call.
-                        attemptSocket.ReplaceHandleIfNecessaryAfterFailedConnect();
+
+                        attemptSocket2?.ReplaceHandleIfNecessaryAfterFailedConnect();
 
                         // Reconfigure the internal event args for the new address.
-                        if (internalArgs.RemoteEndPoint is IPEndPoint existing)
-                        {
-                            existing.Address = address;
-                            Debug.Assert(existing.Port == port);
-                        }
-                        else
+                        if (address != null)
                         {
-                            internalArgs.RemoteEndPoint = new IPEndPoint(address, port);
+                            Debug.Assert(attemptSocket != null);
+                            attemptSocket.ReplaceHandleIfNecessaryAfterFailedConnect();
+                            if (internalArgs.RemoteEndPoint is IPEndPoint existing)
+                            {
+                                existing.Address = address;
+                                Debug.Assert(existing.Port == port);
+                            }
+                            else
+                            {
+                                internalArgs.RemoteEndPoint = new IPEndPoint(address, port);
+                            }
                         }
 
                         // Issue the connect.  If it pends, wait for it to complete.
-                        if (attemptSocket.ConnectAsync(internalArgs))
+                        bool? result = attemptSocket?.ConnectAsync(internalArgs);
```

Comment:
If my previous assumption about `address == null` meaning we should terminate is true, it looks like `attemptSocket` should be never null here.